jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+app.gen() is a "video streaming generator function" which in turn "yield"s frames (in what looks like a MIME-like format) to the caller. I am not clear how video_feed behaves, though.
+
+Basic questions **for chas**:
+
+- What is the overall function of app.py? Do I understand correctly that these decorated functions are called in response to messages from the client webpage? In response to those messages, they appear to both perform internal state-affecting actions (i.e. run the synchronization algorithms, in some cases), *and also* then return the actual HTTP source(*) that should now be displayed in the webpage. Am I understanding that correctly? What is the deal with the "session" object? Could really do with a comment at the top of the module to explain what's going on with all of this.                (*) Oh, or sometimes raw video images apparently, in video_feed()...
+
+- **IMPORTANT:** When app.run() refers to "live", is that what actually gets run for live sync with a microscope, fish and everything? Because, for instance, it calls through to **emulate**_get_period...!? I don't know if this is because, as you said, the app code needs updating, but I can't see how it is meaningfully interacting with cli in the case of **live** frames, and I can't quite picture how that even *should* work. It seems like the gen() function should be being used somehow, but that's not being used except in video_feed().
+
+- base_camera.py needs a similar comment at the top of the module explaining what's going on - and in particular what the relationship is between a "client" and a "thread", explanation of the role of "ident" (which is a unique identifier associated with each thread), etc. Are these concurrent threads or [I suspect] sequential threads? Really needs an explanation of the thread architecture here, and the role of "yield" in terms how of the threads interact.
+
+- **IMPORTANT:** the same questions apply to the Flask app. As per my comment in app.py, I am starting to suspect that this could be running concurrent threads. That needs careful thought and documenting, because it gets very complicated.
+
+- More generally, *as far as I can see*, there doesn't seem to be to be a critical need to be using [sequential] threads here at all (though it doesn't do any harm). Out of interest, is this code just adapted from some online example? I suspect it makes a lot more sense in the case where there is *more than one* client involved (but I don't think that's really an issue for us in reality).
+
+**Out of interest:** I find the code formatting [newline after "(", and ")" on a line of its own, sometimes doing this in rather superfluous contexts] really weird. Is this an established coding style, or just what Alex does!? It's certainly one way to keep to line-width limits, but not the one I would choose - and line limits are violated elsewhere! Extreme example:
```

Comment:
I use the black autoformatter, which does do this a lot... however, your extreme example looked human made?

Reply:
I couldn't say if it's human-made or not, but I copy/pasted it from cli.py!

Reply:
Probably it's black-made but human edited in such a way that black can't un-make it does happen some times.

Reply:
This is something that I do which I believe is a coding style I've copied from Chas - I may have also just read it somewhere.